woocommerce/woocommerce-ios
# Background product image upload: show upload failure notice with an action to view product details

Part of #7021 


### Description


When an image upload fails after the merchant leaves the product form, we want to show an in-app notice about the failure with an action to view product details. It was tricky to only notify the image upload updates when the merchant is **not** in the product form, since we currently show an error alert in the product form when an error occurs. We could also show an in-app notice in the product form, but we still want to customize the notice so that we don't show the action to view product details. I ended up adding two methods `stopEmittingStatusUpdates` and `startEmittingStatusUpdates` to pause the emission of a new `statusUpdates` publisher, when the product form view controller is shown/dismissed. If you have other ideas to implement this, please feel free to share!

Because the image upload error could come from any connected store, I decided to subscribe to the status updates in `MainTabBarController` to present the product (`ProductLoaderViewController`) from the products tab navigation controller after tapping the notice. I also chose to present the product instead of pushing the view controller so that the merchant can come back to any screen they were at (e.g. order or product form) if it's from the same store.

The changes still went over 500 diffs, most of them are on unit tests 🙇🏻‍♀️

### Testing instructions


One way to test image upload failure is by updating the [media remote endpoint](https://github.com/woocommerce/woocommerce-ios/blob/0eae2a783db5f68577d613bf58b42cdda73e744f/Networking/Networking/Remote/MediaRemote.swift#L113) to an invalid one like adding some extra character at the end. You can also put `sleep(numberOfSeconds)` to simulate a longer response time.

Please note that there is a known issue (another subtask of #7021) where an error alert is shown every time you visit the product images screen after an upload failure.

#### Image upload failure outside of product form

- Launch the app
- In the product form, add one or more images
- Tap "Save" to save the product while the images are still being uploaded --> the product is saved remotely first, and there is no "Save" button anymore when the images are still being uploaded
- Leave the product form quickly, and feel free to do anything in the app including switching stores --> when an image fails to upload, an in-app notice (snackbar) should be shown with an action "View" (if you are viewing a modal when the failure occurs, the notice is shown after dismissing the modal)
- Tap on "View" to view product details --> it should switch to the original store if needed, and present the product form

#### Image upload failure within product form (production behavior)

- Launch the app
- In the product form, add one or more images
- Tap "Save" to save the product while the images are still being uploaded --> the product is saved remotely first, and there is no "Save" button anymore when the images are still being uploaded
- Stay in the product form and feel free to navigate within the form --> when an image fails to upload, an error alert should be shown like in production

### Screenshots



https://user-images.githubusercontent.com/1945542/175632039-7b0d2ded-aa96-46d3-bebb-d5de6b0dbead.mov


---
